### DISCUSSION TITLE
Is there a way to define sub projects as intransitive dependencies in a multi-project build?

### DISCUSSION THREAD

**Discussion by tahakonuk** [sbt/sbt#6236] [Q&A] [answered]
Hi,
Say we have following lines in our `build.sbt`

``` 
lazy val root = (project in file("."))
  .aggregate(
    model,
    service,
    gui
  )
  .settings(...)
  )

lazy val model = (project in file("model"))
  .settings(...)

lazy val service = (project in file("service"))
  .settings(...)
  .dependsOn(model)

lazy val gui = (project in file("gui"))
  .settings(...)
  .dependsOn(service)
```

Is there a way to hide `model` from `gui` completely? By completely I mean hiding all dependencies of `model` and every entity defined in `model` project. 
For library dependencies I can use exclusions or `intransitive()` definitions such as;

`lazy val catsCore = "org.typelevel" %% "cats-core" % "2.2.0" intransitive()`

But I can't do that for my local projects.

**Comment by eed3si9n** [comment, marked as answer]
The inter-project dependencies created by aggregating `Compile / exportedProducts` to `Compile / internalDependencyClasspath`. You can do something like the following to remove it from the compilation classpath.

```scala
ThisBuild / organization := "com.example"
ThisBuild / scalaVersion := "2.13.4"

lazy val root = (project in file("."))
  .aggregate(
    model,
    service,
    gui
  )

lazy val model = (project in file("model"))
  .settings(
    name := "model"
  )

lazy val service = (project in file("service"))
  .dependsOn(model)

lazy val gui = (project in file("gui"))
  .dependsOn(service)
  .settings(
    Compile / internalDependencyClasspath --= (model / Compile / exportedProducts).value
  )
```

**Reply by tahakonuk** [reply]
Thank you for quick response. I tried your solution and it definitely works when I run `sbt compile` on the `root`. 

The way I tested is that I have a class in `gui` that uses a class from `model`. So when I compile the project using sbt, compilation for the `gui` fails. Metals(bloop) however ignores it and compiles everything without an error.

Edit for future reference: I was able to switch to sbt build server by following [metals documentation](https://scalameta.org/metals/docs/build-tools/sbt.html#sbt-build-server). Now I get the same behaviour as I run sbt manually via metals build commands.
